src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java

```diff
@@ -188,4 +140,115 @@ public boolean canUnarchiveArtifact() {
+                        // when we add support for private non-ECR registries, this can be expanded to retrieve
+                        // credentials for them on case by case basis
+                        if (image.getRegistry().isEcrRegistry()) {
+                            // Get auth token for ECR, which represents ECR registry credentials
+                            Registry.Credentials credentials =
+                                    ecrAccessor.getCredentials(image.getRegistry().getRegistryId());
+                            image.getRegistry().setCredentials(credentials);
+                            credentialRefreshNeeded.set(false);
+                        }
+                        return null;
+                    }, "get-ecr-auth-token", logger);
+                } catch (InterruptedException e) {
+                    throw e;
+                } catch (Exception e) {
+                    throw new PackageDownloadException(getErrorString("Failed to get auth token for docker login"), e);
+                }
+
+                // Login to registry
+                // TODO: [P44950158]: Avoid logging into registries which might already have been logged in previously
+                //  with the same and valid credentials by maintaining a cache across artifacts and deployments
+                run(() -> {
+                    if (credentialsUsable(image)) {
+                        dockerClient.login(image.getRegistry());
+                    } else {
+                        // Credentials have expired, re-fetch and login again
+                        logger.atInfo().kv("registry-endpoint", image.getRegistry().getEndpoint())
+                                .log("Registry credentials have expired,"
+                                        + "fetching fresh credentials and logging in again");
+                        credentialRefreshNeeded.set(true);
+                    }
+                    return null;
+                }, "docker-login", "Failed to login to docker registry");
+            }
+
+            // Don't even attempt to pull images if credentials are expired, first try to refresh credentials
+            if (credentialRefreshNeeded.get()) {
+                continue;
+            }
+
+            // Docker pull
+            run(() -> {
+                if (credentialsUsable(image)) {
+                    dockerClient.pullImage(image);
+                } else {
+                    // Credentials have expired, re-fetch and login again
+                    logger.atInfo().kv("registry-endpoint", image.getRegistry().getEndpoint())
+                            .log("Registry credentials have expired, fetching fresh credentials and logging in again");
+                    credentialRefreshNeeded.set(true);
+                }
+                return null;
+            }, "docker-pull-image", "Failed to download docker image");
+        } while (credentialRefreshNeeded.get());
```

Comment:
why do we need this retry mechanism? if we always gets the credentials before making a pull, is there an edge case where the creds will expire after fetching but before pulling the image?

Reply:
Credentials can expire if the device is offline for longer than 12 hours and the downloader waits for connection for executing login/pull

Reply:
The pull is retried only 30 times, if the device is offline for 12 hours wouldnt all the retries be exhausted?

Reply:
It's retried 30 times when we can't tell if the error can be due to connection problems, but indefinitely when we know the device is offline, that's why I have the `runWithConnectionError` stuff